source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.h

```diff
@@ -62,23 +81,64 @@ class SPIFFEValidator : public CertValidator {
   X509_STORE* getTrustBundleStore(X509* leaf_cert);
   static std::string extractTrustDomain(const std::string& san);
   static bool certificatePrecheck(X509* leaf_cert);
-  absl::flat_hash_map<std::string, X509StorePtr>& trustBundleStores() {
-    return trust_bundle_stores_;
+  std::shared_ptr<SpiffeData> getSpiffeData() const {
+    if (tls_) {
+      return tls_->get()->getSpiffeData();
+    }
+    return spiffe_data_;
   };
-
   bool matchSubjectAltName(X509& leaf_cert);
 
 private:
   bool verifyCertChainUsingTrustBundleStore(X509& leaf_cert, STACK_OF(X509)* cert_chain,
                                             X509_VERIFY_PARAM* verify_param,
                                             std::string& error_details);
 
+  void initializeCertificateRefresh(Server::Configuration::CommonFactoryContext& context);
+  std::shared_ptr<SpiffeData> parseTrustBundles(std::string trust_bundles_str);
+
+  class ThreadLocalSpiffeState : public Envoy::ThreadLocal::ThreadLocalObject {
+  public:
+    std::shared_ptr<SpiffeData> getSpiffeData() const { return spiffe_data_; }
+    void updateSpiffeData(std::shared_ptr<SpiffeData> new_data) {
+      ENVOY_LOG(debug, "updating spiffe data");
+      spiffe_data_ = new_data;
+    }
+
+  private:
+    std::shared_ptr<SpiffeData> spiffe_data_;
+  };
+
+  void postSpiffeDataUpdate(std::shared_ptr<SpiffeData> new_spiffe_data) {
+    ENVOY_LOG(debug, "Posting new SPIFFE data update to main thread dispatcher");
+    main_thread_dispatcher_.post([this, new_spiffe_data]() {
+      ENVOY_LOG(debug, "Updating spiffe_data_ for all threads");
+      tls_->runOnAllThreads([new_spiffe_data](OptRef<ThreadLocalSpiffeState> obj) {
+        ENVOY_LOG(debug, "loading new spiffe data");
+        obj->updateSpiffeData(new_spiffe_data);
+      });
+    });
+  };
+
```

Comment:
As I said before, I think this is unecessary. Or if you have any reason to use this?

Reply:
Using the dispatcher post was causing some race condition crashing at startup.  Blocking with a direct call to runOnAllThreads was how I worked around this.  Still use dispatcher post for dynamic updates.

Reply:
I mean why you didn't call the `tls_->runOnAllTreads`? Why we wrap it with another post?

Reply:
/wait-any